Userland/Libraries/LibC/pwd.cpp

```diff
@@ -166,91 +160,38 @@ int getpwent_r(struct passwd* passwd_buf, char* buffer, size_t buffer_size, stru
     }
 }
 
-static void construct_pwd(struct passwd* pwd, char* buf, struct passwd** result)
+// https://pubs.opengroup.org/onlinepubs/9699919799/functions/getpwnam.html
+int getpwnam_r(char const* name, struct passwd* pwd, char* buffer, size_t bufsize, struct passwd** result)
 {
-    auto* buf_name = &buf[0];
-    auto* buf_passwd = &buf[s_name.length() + 1];
-    auto* buf_gecos = &buf[s_name.length() + 1 + s_gecos.length() + 1];
-    auto* buf_dir = &buf[s_gecos.length() + 1 + s_name.length() + 1 + s_gecos.length() + 1];
-    auto* buf_shell = &buf[s_dir.length() + 1 + s_gecos.length() + 1 + s_name.length() + 1 + s_gecos.length() + 1];
-
-    bool ok = true;
-    ok = ok && s_name.copy_characters_to_buffer(buf_name, s_name.length() + 1);
-    ok = ok && s_passwd.copy_characters_to_buffer(buf_passwd, s_passwd.length() + 1);
-    ok = ok && s_gecos.copy_characters_to_buffer(buf_gecos, s_gecos.length() + 1);
-    ok = ok && s_dir.copy_characters_to_buffer(buf_dir, s_dir.length() + 1);
-    ok = ok && s_shell.copy_characters_to_buffer(buf_shell, s_shell.length() + 1);
-
-    VERIFY(ok);
-
-    *result = pwd;
-    pwd->pw_name = buf_name;
-    pwd->pw_passwd = buf_passwd;
-    pwd->pw_gecos = buf_gecos;
-    pwd->pw_dir = buf_dir;
-    pwd->pw_shell = buf_shell;
-}
-
-int getpwnam_r(char const* name, struct passwd* pwd, char* buf, size_t buflen, struct passwd** result)
-{
-    // FIXME: This is a HACK!
-    TemporaryChange name_change { s_name, {} };
-    TemporaryChange passwd_change { s_passwd, {} };
-    TemporaryChange gecos_change { s_gecos, {} };
-    TemporaryChange dir_change { s_dir, {} };
-    TemporaryChange shell_change { s_shell, {} };
-
     setpwent();
-    bool found = false;
-    while (auto* pw = getpwent()) {
-        if (!strcmp(pw->pw_name, name)) {
-            found = true;
-            break;
+    while (getpwent_r(pwd, buffer, bufsize, result) == 0) {
```

Comment:
We should maybe add a FIXME here (and same below) to use `fgetpwent_r` once we implement it, so we don't forget? As you said in the video, this is better than it was before but it's still not thread safe.

Reply:
Even if we had `fgetpwent_r()`, we still have to use `getpwent_r()` here since this function doesn't receive a `FILE*`.

These APIs are pretty weird, and we don't use them ourselves in SerenityOS. They're just provided for port compatibility, so we needn't be too worried about them, as long as they behave the same as on other platforms :)